### ISSUE TITLE
Reorganize Mapping classes

### ISSUE THREAD

**Issue by yguclu** [pyccel/psydac#189]
Psydac has an abstract base class `Mapping`, in file [psydac/psydac/mapping/basic.py](https://github.com/pyccel/psydac/blob/devel/psydac/mapping/basic.py), which defines a simple interface for evaluation of the mapping and its derivatives, for very different kinds of mappings.

This class is not used anymore. In the past it was subclassed by:

   - `AnalyticalMapping`, which
      - **does not exist anymore**, and is (partially) replaced by Sympde's `CallableMapping`
      - was also an abstract class, subclassed by concrete mappings (e.g. `PolarMapping`, `TargetMapping`, etc.)
      - contained a `SymbolicMapping` object (made of pure Sympy expressions)
      - stored any free parameters in a dictionary of `(symbol, value)` pairs

   - `SplineMapping`, which
      - **does not subclass `Mapping` anymore**
      - represents a discrete spline mapping
 
   - `NurbsMapping`, which
      - **does not subclass `Mapping` anymore**
      - represents a discrete NURBS mapping

The current situation is confusing and sub-optimal. We should (probably) make sure that all callable mappings in both Psydac and Sympde subclass a common abstract base class, which defines a unique interface across the two libraries.

**Comment by anlavandier** [comment]
I agree. I think that what makes the most sense is to have a `CallableMapping` class that is subclassed by `Spline/NurbsMapping`. This `CallableMapping` class would have a `.__new__()` that accept SymPDE Mappings and does with them something very similar to `get_callable_mapping`. That way, we could also use accelerated evaluation methods for the SymPDE mappings.

**Comment by yguclu** [comment]
@anlavandier You should have a look at the `CallableMapping` class in SymPDE, as it does exactly what you describe. [`Mapping.get_callable_mapping()`](https://github.com/pyccel/sympde/blob/7e2a5622bb2eb3bf4a592db9113b1c9f9fdd4dcf/sympde/topology/mapping.py#L327-L332) is essentially passing `self` to [`CallableMapping.__init__()`](https://github.com/pyccel/sympde/blob/master/sympde/topology/callable_mapping.py).

I am not sure of what you mean by "accelerated evaluation methods for the SymPDE mappings". `CallableMapping` uses `lambdify` to get functions which can be evaluated at Numpy speed. If Pyccel could provide an equivalent `lambdify` method, we could generate Fortran/C code instead of Python.

**Comment by anlavandier** [comment]
The problem that might arise is about having `SplineMapping` be a subclass of those `CallableMapping`. Maybe we should have the discrete mappings stay as is and add a `get_callable_mapping()` method  for consistency and to not have to change the `SplineMapping` class.
